### ISSUE TITLE
BZPOPMIN will hang for ever when a short disconnection happens and the port is different from 6379

### ISSUE THREAD

**Issue by manast** [redis/ioredis#1888]
This one was pretty weird but quite serious. If you issue a BZPOPMIN command, and there is a small disconnection while the command is blocking, then it will stay blocking forever, effectively hanging the app that uses the command. 

Now, one of the *insane* things about this issue is that to reproduce it you must use a port different than the standard 6379, and also use docker network disconnect, it is not enough to just stop the docker container running Redis:

Try this code:
bug.mjs:
```ts
import Redis from "ioredis";

const connection = new Redis({
    host: "localhost",
    port: 6380,
});

connection.on("connect", () => console.log("Redis connected!"));
connection.on("ready", () => console.log("Redis ready!"));
connection.on("error", (err) => console.error("Redis error:", err));
connection.on("end", () => console.log("Redis connection ended"));
connection.on("reconnecting", () => console.log("Redis reconnecting..."));

async function test() {
    console.log("Gonna issue a BZPOPMIN command...")
    const result = await connection.bzpopmin("key", 0)
    console.log("Issued BZPOPMIN command!", result)
}

test().catch(console.error);
```

docker-compose.yml:
```yaml
version: "3.7"
services:
  redis:
    image: redis:alpine
    ports:
      - "6380:6379"
```

Start the docker container with:
```
docker-compose up
```

Run the test with:
```sh
node bug.mjs
```
Results in:
```sh
Gonna issue a BZPOPMIN command...
Redis connected!
Redis ready!
```

Now in a different terminal, and depending on where you run the docker (I had my docker compose on directory ioredis-econnreset:
```sh
docker network disconnect ioredis-econnreset_default ioredis-econnreset-redis-1
sleep 3
docker network connect ioredis-econnreset_default ioredis-econnreset-redis-1
```

The Redis network was disconnected and connected again after 3 seconds. Open a Redis cli:
```sh
redis-cli -p 6380
> 127.0.0.1:6380> zadd key 10 test
(integer) 1
127.0.0.1:6380>
```

The terminal with the node app will stay the same, with no error or nothing. If you restart the app and without disconnecting and reconnecting you try again to ZADD to the key you will get instead:
```sh
Gonna issue a BZPOPMIN command...
Redis connected!
Redis ready!
Issued BZPOPMIN command! [ 'key', 'test', '10' ]
```

Let me know if you need more information.

**Comment by lukas-becker0** [comment]
Hi @manast,

I can not reproduce this with ioredis `5.4.1`, node `20.14.0`, the latest redis alpine version (`7.2.5`) and docker `26.1.3` (colima on macos).

> The terminal with the node app will stay the same, with no error or nothing.

I do get connection errors and reconnecting output as expected when disconnecting 
and after a few seconds reconnecting the docker network like you described ( I also tried it with less time between the disconnect and connect docker commands (1 second) and more than 3 seconds >=10 seconds but the result is the same).

I always get the `Issued BZPOPMIN command! [ 'key', 'test', '10' ]` 
output in the apps terminal as expected even without restarting the app.

<img width="802" alt="image" src="https://github.com/redis/ioredis/assets/117725216/4774762b-297f-4b61-83ed-3b648658fde8">

**Comment by manast** [comment]
Sorry to ping about this issue again, but I can still reproduce it and it is causing some issues to BullMQ users... Could somebody from the team contact me so that we can do a zoom call and I can show them how to reproduce it? I am actually in a different computer than when I reported this issue the first time, and I can still reproduce it.

**Comment by manast** [comment]
This is indeed an edge case and probably difficult to fix, but it is a real issue and is causing major problems in production, so I think it should not be overlooked...

**Comment by manast** [comment]
And it really happens only when using a different port than the default port, so if I have to guess, somewhere deep in the retry functionality or connection error functionality it is somehow loosing the configured port value and is using the default port value and not properly reconnecting (nor emitting any error either).

**Comment by manast** [comment]
I have been making more investigations, and it seems that the issue will only be reproduced using a timeout of 0. If we use a timeout of 5 for example, then after 5 seconds, if the connection to Redis has been interrupted (by using docker network disconnect, which normally gets unnoticed during the blocking commands) it will detect the disconnection when the timeout has passed. So this would mean that as long as we have some timeout, and not wait indefinitely with timeout 0, it will work. This is what Gemini 3 gave me as an explanation: "The reason you don't see a connection error on port 6380 is that disconnecting the Docker network simulates a "silent" cable cut. In TCP, if the client is just waiting for data (like in [bzpopmin](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)) and not sending anything, it has no way of knowing the link is dead until the OS TCP stack times out (which can take minutes or hours) or Keepalive probes fail.

Port 6380 likely "works" because of how Docker Desktop on macOS handles the default port forwarding (via vpnkit or a user-mode proxy) which might be actively sending a RST packet to the host when the backend network changes, whereas the custom port mapping on 6380 behaves like a true "black hole" partition."

**Comment by manast** [comment]
I am sorry, I spoke too soon. It will not work regardless of having a timeout or not. The reason it was working in my tests is because Gemini actually modified IORedis and added an external timeout that disconnects if Redis does not respond in that timeout interval.

So the bug seems to be clear now, what is happening is that when issuing blocking commands such as BZPOPMIN, if during the timeout interval there is a certain type of disconnection, IORedis will not detect it and since it is 100% relying on redis returning from that command, and this never happens as the "cable was cut off" so to speak, meaning that Redis may have returned the response but the response was never received by IORedis, the command will effectively be hanging forever... this is a problem.

**Comment by PavelPashov** [comment]
Hey @manast, thank you for looking into this. I’ll add it to my list to review.

**Comment by manast** [comment]
Please take a look at the PR I submitted. Note that I am not familiar with IORedis codebase, this was mostly developed by Codex.

**Comment by thefrana** [comment]
Thanks @manast for looking into this. We are long time users of BullMQ and we had to manually apply some patches because we regularly encountered this issue. Weird thing was it happened only in our Kubernetes cluster and never on dev.

**Comment by manast** [comment]
@thefrana yes because it requires a very particular kind of disconnect where basically the underlaying TCP connection does not detect the disconnection so it stays pseudo connected for ever. By the way, this also implies that having timeout 0 is a bad design decision as it would potentially lead to a command issued that will never resolve. I am not good enough in low level TCP coding so I do not know if this is something NodeJS specific or would be an issue for all Redis clients though...

**Comment by thefrana** [comment]
Yes, we come up with something similar in the end. Is there anything we can do? The result is it keeps reconnecting upon timeout (in our case 10s).

**Comment by PavelPashov** [comment]
@manast & @thefrana  please take a look at #2051 and let me know your thoughts. Additionally, let me know if you would be interested in reviewing a beta version that includes this functionality.
